cmd/zb/perf.go

```diff
@@ -622,6 +625,215 @@ func PushChunkStreamed(workdir, url, auth, trepo string, requests int,
 	return nil
 }
 
+func PullMonolithStreamed(workdir, url, auth, trepo string, requests int,
+	config testConfig, statsCh chan statsRecord) error {
+	client := resty.New()
+
+	if auth != "" {
+		creds := strings.Split(auth, ":")
+		client.SetBasicAuth(creds[0], creds[1])
+	}
+
+	// Push blob for read it later
+	var statusCode int
+
+	ruid, err := uuid.NewUUID()
+	if err != nil {
+		return err
+	}
+
+	var repo string
+
+	if trepo != "" {
+		repo = trepo + "/" + ruid.String()
+	} else {
+		repo = ruid.String()
+	}
+
+	// upload blob
+	resp, err := resty.R().Post(fmt.Sprintf("%s/v2/%s/blobs/uploads/", url, repo))
+	if err != nil {
+		return err
+	}
+
+	// request specific check
+	statusCode = resp.StatusCode()
+	if statusCode != http.StatusAccepted {
+		return errors.ErrUnknownCode
+	}
+
+	loc := location(url, resp)
+
+	size := config.size
+	blob := path.Join(workdir, fmt.Sprintf("%d.blob", size))
+
+	fhandle, err := os.OpenFile(blob, os.O_RDONLY, defaultFilePerms)
+	if err != nil {
+		return err
+	}
+
+	defer fhandle.Close()
+
+	// stream the entire blob
+	digest := blobHash[blob]
+
+	resp, err = client.R().
+		SetContentLength(true).
+		SetQueryParam("digest", digest.String()).
+		SetHeader("Content-Length", fmt.Sprintf("%d", size)).
+		SetHeader("Content-Type", "application/octet-stream").SetBody(fhandle).Put(loc)
+
+	if err != nil {
+		return err
+	}
+
+	// request specific check
+	statusCode = resp.StatusCode()
+	if statusCode != http.StatusCreated {
+		return errors.ErrUnknownCode
+	}
+
+	// create a manifest
+	manifest := ispec.Manifest{
+		Versioned: imeta.Versioned{
+			SchemaVersion: defaultSchemaVersion,
+		},
+		Config: ispec.Descriptor{
+			Digest: digest,
+			Size:   int64(size),
+		},
+		Layers: []ispec.Descriptor{
+			{
+				MediaType: "application/vnd.oci.image.layer.v1.tar",
+				Digest:    digest,
+				Size:      int64(size),
+			},
+		},
+	}
+
+	content, err := json.Marshal(manifest)
+	if err != nil {
+		return err
+	}
+
+	digest = godigest.FromBytes(content)
+
+	resp, err = resty.R().SetHeader("Content-Type", "application/vnd.oci.image.manifest.v1+json").
+		SetBody(content).Put(fmt.Sprintf("%s/v2/%s/manifests/%s", url, repo, digest.String()))
+
+	if err != nil {
+		return err
+	}
+
+	// request specific check
+	statusCode = resp.StatusCode()
+	if statusCode != http.StatusCreated {
+		return errors.ErrUnknownCode
+	}
+
+	manifestLoc := location(url, resp)
+
+	// download image
+	for count := 0; count < requests; count++ {
+		func() {
+			start := time.Now()
+
+			var isConnFail, isErr bool
+
+			var statusCode int
+
+			var latency time.Duration
+
+			defer func() {
+				// send a stats record
+				statsCh <- statsRecord{
+					latency:    latency,
+					statusCode: statusCode,
+					isConnFail: isConnFail,
+					isErr:      isErr,
+				}
+			}()
+
```

Comment:
Some steps have been skipped.
Start with HEAD manifest (to check manifest/tag present), GET manifest, HEAD each blob and then GET each blob (this you have currently)

Reply:
@roxanaN, if you like, you can move some code into pull.go and push.go - your choice